patch: serialization/deserialization of ConsequenceQuery

Ant-hem opened the pull request:
| Q                 | A
| ----------------- | ----------
| Bug fix?          | yes
| New feature?      | no
| BC breaks?        | no     
| Related Issue     | Fix #662 
| Need Doc update   | no


## Problem
The custom serializer was not handling all the cases for the payload below.

Example:

```json
// query string
"query": "some query string"

// remove attribute (deprecated)
"query": {"remove": ["query1", "query2"]}

// edits attribute
"query": {
   "edits": [
     { "type": "remove", "delete": "old" },
     { "type": "replace", "delete": "new", "insert": "newer" }
   ]
}}
```

## Fix

I added a `string` property in `ConsequenceQuery` and modified the serializer to allow `ConsequenceQuery` to deserialize and serializer Query as string. This required to overload `setQuery` and `getQuery` from `SearchParameters`

nunomaduro commented:
@Ant-hem Thanks for this. There any breaking change concerning what we had before?  Or his just adding a new feature for a particular situation that wasn't possible before?

Ant-hem committed:
patch: serialization/deserialization of ConsequenceQuery

Fixed: Custom serializer to handle polymorphism of "query" attribute
in ConsequenceQuery

Example:

```json
// query string
"query": "some query string"

// remove attribute (deprecated)
"query": {"remove": ["query1", "query2"]}

// edits attribute
"query": {
   "edits": [
   { "type": "remove", "delete": "old" },
   { "type": "replace", "delete": "new", "insert": "newer" }
   ]
}}
```json

Ant-hem commented:
> @Ant-hem Thanks for this. There any breaking change concerning what we had before? Or his just adding a new feature for a particular situation that wasn't possible before?

It's a backwards compatible bug fix. To give more context:  before the patch, using `setQuery` (in `Rule.ConsequenceQuery`) would have produced an empty payload. Now it's writing the correct one. So I had to modify both serializer and deserializer.

nunomaduro commented:
Great! Let's wait for @aseure to review the code.

aseure commented on algoliasearch-core/src/main/java/com/algolia/search/models/rules/ConsequenceQuery.java:
```diff
@@ -31,29 +37,118 @@ public ConsequenceQuery setEdits(List<Edit> edits) {
+     * https://www.algolia.com/doc/api-reference/api-methods/save-rule/#method-param-query
+     * */
+    if (!AlgoliaUtils.isNullOrEmptyWhiteSpace(value.getQueryString())) {
+      gen.writeString(value.getQueryString());
```
You may want to migrate `value.getQueryString()` in its own temporary: it's probably negligible in term of performances but for readability, it would help at least.

Ant-hem replied:
> You may want to migrate `value.getQueryString()` in its own temporary: it's probably negligible in term of performances but for readability, it would help at least.

Could you elaborate? i don't get it.

aseure replied:
I was suggesting something like this:

```java
String query = value.getQueryString();

 if (!AlgoliaUtils.isNullOrEmptyWhiteSpace(query)) {
      gen.writeString(query);
} else {
     // ...
}
```

Ant-hem committed:
fix: remove null check in setter

aseure reviewed (approved):
Approving the change regarding the override of the `ConsequenceQuery` when calling `.setQuery()`.
